lessons/advanced/gen-stage.md

```diff
@@ -0,0 +1,261 @@
+
+In this example we have three stages: `A` a producer, `B` a producer-consumer, and `C` a consumer.  `A` produces a value which is consumed by `B`, `B` performs some work and returns a new value which is received by our consumer `C`; the role of our stage is important as we'll see in the next section.
+
+While our example is 1-to-1 producer-to-consumer it's possible to both have multiple producers and multiple consumers at any given stage.
+
+To better illustrate these concepts we'll be constructing a pipeline with GenStage but first let's explore the roles that GenStage relies upon a bit further.
+
+## Consumers and Producers
+
+As we've read, the role we give our stage is important.  The GenStage specification recognizes three roles:
+
++ `:producer` — A source.  Producers wait for demand from consumers and respond with the requested events.
+
++ `:producer_consumer` — Both a source and a sink.  Producer-consumers can respond to demand from other consumers as well as request events from producers.
+
++ `:consumer` — A sink.  A consumer requests and receives data from producers.
+
+Notice that our producers _wait_ for demand?  With GenStage our consumers send demand upstream and process the data from our producer.  This faciliates a mechanism known as back-pressure.  Back-pressure puts the onerous on the producer to not over-pressure when consumers are busy.
+
+Now that we've covered the roles within GenStage let's start on our application.
+
+## Getting Started
+
+In this example we'll be constructing a GenStage application that omits numbers, sort out the prime numbers, and finally print them.
+
+For our application we'll use all three GenStage roles.  Our producer will be responsible for counting and emits numbers.  We'll use a producer-consumer to filter out only the even numbers and later respond to demand from downsteam.  Last we'll build a consumer the remaining numbers for us.
+
+We'll begin by generating a project with a supervision tree:
+
+```shell
+$ mix new genstage_example --sup
+$ cd genstage_example
+```
+
+Let's update our dependencies in `mix.exs` to include `gen_stage`:
+
+```elixir
+  defp deps do
+    [
+      {:gen_stage, "~> 0.7},
+    ]
+  end
+```
+
+We should fetch our dependencies and compile before go much further:
+
+```shell
+$ mix do deps.get, compile
+```
+
+Now we're ready to build our producer!
+
+## Producer
+
+The first step of our GenStage application is creating our producer.  As we discussed before we want to create a producer that emits a constant stream of numbers.  Let's create our producer file:
+
+```shell
+$ mkdir lib/genstage_example
+$ touch lib/genstage_example/producer.ex
+```
+
+Now we can add the code:
+
+```
+defmodule GenstageExample.Producer do
+  alias Experimental.GenStage
+  
+  use GenStage
+  
+  def start_link(initial \\ 0) do
+    GenStage.start_link(__MODULE__, initial, name: __MODULE__)
+  end
+
+  def init(counter), do: {:producer, counter}
+
+  def handle_demand(demand, state) do
```

Comment:
I think here it is worth mentioning what exactly demand is. While it _might_ seem obvious that its some number N, it personally took me a moment to grok this when I was diving in. Maybe we should just mention that it defaults to 1000, and mention how that generates this list here, and that by sending a nonempty list we will force the consumer to act.